### PR TITLE
MOTECH-2581: Upgrade OpenMRS model

### PR DESCRIPTION
**Pull request by pkornowski**

### CODE REVIEW

File: openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java [hunk 1]
```diff
@@ -135,6 +136,63 @@ public boolean equals(Object obj) {
         return Objects.equals(this.uuid, other.uuid) && Objects.equals(this.display, other.display) &&
                 Objects.equals(this.location, other.location) && Objects.equals(this.encounterType, other.encounterType) &&
                 Objects.equals(this.encounterDatetime, other.encounterDatetime) && Objects.equals(this.patient, other.patient) &&
-                Objects.equals(this.provider, other.provider) && Objects.equals(this.obs, other.obs);
+                Objects.equals(this.encounterProviders, other.encounterProviders) && Objects.equals(this.obs, other.obs);
+    }
+
```
Comment (matkwiatkowski) [inline openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java:141]: add javadoc like in [Observation.java](https://github.com/motech/modules/blob/master/openmrs/src/main/java/org/motechproject/openmrs/domain/Observation.java)

File: openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java [hunk 2]
```diff
@@ -135,6 +136,63 @@ public boolean equals(Object obj) {
...
+                encounter.addProperty("provider", src.getEncounterProviders().get(0).getUuid());
+            }
+            if (src.obs != null) {
+                final JsonElement jsonObs = context.serialize(src.getObs());
+                encounter.add("obs", jsonObs);
+            }
+            return encounter;
+        }
+    }
+
```
Comment (matkwiatkowski) [inline openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java:180]: same as above

File: openmrs/src/main/java/org/motechproject/openmrs/resource/impl/EncounterResourceImpl.java [hunk 3]
```diff
@@ -27,32 +30,33 @@ public EncounterResourceImpl(RestOperations restOperations) {
...
     @Override
     public EncounterListResult queryForAllEncountersByPatientId(Config config, String id) {
         String responseJson = getJson(config, "/encounter?patient={id}&v=full", id);
-        
-        EncounterListResult encounterList = (EncounterListResult) JsonUtils.readJson(responseJson, EncounterListResult.class);
-        for(Encounter encounter : encounterList.getResults()) {
-            setProviderFromEncounterProviderList(encounter);
+
+        List<Encounter> encounterList = new ArrayList<>();
+        JsonArray allEncounters = prepareJsonArray(responseJson);
```
Comment (matkwiatkowski) [inline openmrs/src/main/java/org/motechproject/openmrs/resource/impl/EncounterResourceImpl.java]: rename prepareJsonArray function to prepareAllEncountersList and then remove additional allEncounters variable and call that function inside for declaration

**Comment by matkwiatkowski** [comment]
@kmadej looks good for me

File: openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java [hunk 4]
```diff
@@ -135,6 +136,69 @@ public boolean equals(Object obj) {
...
+     * Implementation of the {@link JsonDeserializer} interface for the {@link Encounter} class.
+     */
+    public static class EncounterDeserializer implements JsonDeserializer<Encounter> {
+        @Override
+        public Encounter deserialize(final JsonElement src, final Type typeOfT, final JsonDeserializationContext context) {
+
+            final JsonObject jsonEncounter = src.getAsJsonObject();
+
+            final JsonElement jsonProvider = jsonEncounter.get("provider");
+            Person provider;
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java]: this variable declaration is redundant, do it with assignment in one line

File: openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java [hunk 5]
```diff
@@ -135,6 +136,69 @@ public boolean equals(Object obj) {
...
+            provider = (Person) JsonUtils.readJson(jsonProvider.toString(), Person.class);
+
+            List<Person> encounterProviders = new ArrayList<>();
+            encounterProviders.add(provider);
+
+            Encounter encounter;
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java]: the same as above

File: openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java [hunk 6]
```diff
@@ -135,6 +136,69 @@ public boolean equals(Object obj) {
...
+            encounter = (Encounter) JsonUtils.readJson(src.toString(), Encounter.class);
+            encounter.setEncounterProviders(encounterProviders);
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java]: singletonList(provider)

File: openmrs/src/main/java/org/motechproject/openmrs/resource/impl/EncounterResourceImpl.java [hunk 7]
```diff
@@ -82,26 +86,32 @@ private Gson buildGson() {
...
-            encounter.setProvider(encounter.getEncounterProviders().get(0));
+    private Gson buildGsonWithAdaptersDeserialize() {
+        return new GsonBuilder()
+                .registerTypeAdapter(Encounter.class, new Encounter.EncounterDeserializer())
+                .create();
+    }
+
+    private Encounter checkVersionAndSetEncounter(Config config, String responseJson) {
+        Encounter createdEncounter;
+        if ("1.9".equals(config.getOpenMrsVersion())) {
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/resource/impl/EncounterResourceImpl.java]: use constant here, OPENMRS_V1.9

File: openmrs/src/main/java/org/motechproject/openmrs/resource/impl/EncounterResourceImpl.java [hunk 8]
```diff
@@ -82,26 +86,32 @@ private Gson buildGson() {
...
+            createdEncounter = buildGsonWithAdaptersDeserialize().fromJson(responseJson, Encounter.class);
+        } else {
+            createdEncounter = (Encounter) JsonUtils.readJson(responseJson, Encounter.class);
         }
+        return createdEncounter;
+    }
+
+    private  JsonArray prepareAllEncountersList(String json) {
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/resource/impl/EncounterResourceImpl.java]: additional space

File: openmrs/src/test/java/org/motechproject/openmrs/validation/ConfigValidatorTest.java [hunk 9]
```diff
@@ -20,6 +21,7 @@
     public void shouldPassTheValidationIfAllTheFieldsAreNonEmpty() {
         Config config = new Config();
```
Comment (kmadej) [inline openmrs/src/test/java/org/motechproject/openmrs/validation/ConfigValidatorTest.java]: there is a lot of duplicated code here, please use constructor or a private method for setting all these fields

File: openmrs/src/main/java/org/motechproject/openmrs/resource/impl/EncounterResourceImpl.java [hunk 10]
```diff
@@ -111,6 +113,7 @@ private Encounter checkVersionAndSetEncounter(Config config, String responseJson
     }
 
     private  JsonArray prepareAllEncountersList(String json) {
+
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/resource/impl/EncounterResourceImpl.java]: remove this line, there additional space between private and JsonArray

File: openmrs/src/main/java/org/motechproject/openmrs/resource/EncounterResource.java [hunk 11]
```diff
@@ -5,6 +5,7 @@
 import org.motechproject.openmrs.domain.EncounterListResult;
 import org.motechproject.openmrs.config.Config;
 
+
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/resource/EncounterResource.java]: remove this line

**Comment by kmadej** [comment]
test this please

**Pull request merged**